From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Wed, 27 Sep 2023 18:29:51 +0900
Subject: [PATCH] Plazma: Add some missing Pufferfish configurations

Original license: MIT
Original project: https://github.com/PlazmaMC/PlazmaBukkit

Add Pufferfish DAB support for mobs below:
Armadillo, Camel, CopperGolem, HappyGhast, Nautilus, ZombieNautilus, Sniffer, Zoglin, Breeze, Creaking, PiglinBrute
Resolved https://github.com/pufferfish-gg/Pufferfish/issues/83

diff --git a/net/minecraft/world/entity/animal/armadillo/Armadillo.java b/net/minecraft/world/entity/animal/armadillo/Armadillo.java
--- a/net/minecraft/world/entity/animal/armadillo/Armadillo.java
+++ b/net/minecraft/world/entity/animal/armadillo/Armadillo.java
@@ -187,10 +187,12 @@ public class Armadillo extends Animal {
         return (Brain<Armadillo>)super.getBrain();
     }
 
+    private int behaviorTick; // Leaf - Plazma - Add missing Pufferfish configurations
     @Override
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("armadilloBrain");
+        if (this.behaviorTick++ % this.activatedPriority == 0) // Leaf - Plazma - Add missing Pufferfish configurations
         this.getBrain().tick(level, this);
         profiler.pop();
         profiler.push("armadilloActivityUpdate");
diff --git a/net/minecraft/world/entity/animal/camel/Camel.java b/net/minecraft/world/entity/animal/camel/Camel.java
--- a/net/minecraft/world/entity/animal/camel/Camel.java
+++ b/net/minecraft/world/entity/animal/camel/Camel.java
@@ -199,11 +199,13 @@ public class Camel extends AbstractHorse {
             : (this.isBaby() ? BABY_STANDING_DIMENSIONS : super.getDefaultDimensions(pose));
     }
 
+    private int behaviorTick = 0; // Leaf - Plazma - Add missing Pufferfish configurations
     @Override
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("camelBrain");
         Brain<Camel> brain = this.getBrain();
+        if (this.behaviorTick++ % this.activatedPriority == 0) // Leaf - Plazma - Add missing Pufferfish configurations
         brain.tick(level, this);
         profiler.pop();
         profiler.push("camelActivityUpdate");
diff --git a/net/minecraft/world/entity/animal/golem/CopperGolem.java b/net/minecraft/world/entity/animal/golem/CopperGolem.java
--- a/net/minecraft/world/entity/animal/golem/CopperGolem.java
+++ b/net/minecraft/world/entity/animal/golem/CopperGolem.java
@@ -238,11 +238,12 @@ public class CopperGolem extends AbstractGolem implements ContainerUser, Shearab
         this.setSummoner(input.read("Purpur.Summoner", net.minecraft.core.UUIDUtil.CODEC).orElse(null)); // Purpur - Summoner API
     }
 
+    private int behaviorTick; // Leaf - Plazma - Add missing Pufferfish configurations
     @Override
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("copperGolemBrain");
-        if (getRider() == null || !this.isControllable()) // Purpur - only use brain if no rider
+        if ((getRider() == null || !this.isControllable()) && this.behaviorTick++ % this.activatedPriority == 0) // Purpur - only use brain if no rider // Leaf - Plazma - Add missing Pufferfish configurations
         this.getBrain().tick(level, this);
         profiler.pop();
         profiler.push("copperGolemActivityUpdate");
diff --git a/net/minecraft/world/entity/animal/happyghast/HappyGhast.java b/net/minecraft/world/entity/animal/happyghast/HappyGhast.java
--- a/net/minecraft/world/entity/animal/happyghast/HappyGhast.java
+++ b/net/minecraft/world/entity/animal/happyghast/HappyGhast.java
@@ -442,11 +442,13 @@ public class HappyGhast extends Animal {
         return (Brain<HappyGhast>)super.getBrain();
     }
 
+    private int behaviorTick; // Leaf - Plazma - Add missing Pufferfish configurations
     @Override
     protected void customServerAiStep(final ServerLevel level) {
         if (this.isBaby()) {
             ProfilerFiller profiler = Profiler.get();
             profiler.push("happyGhastBrain");
+            if (this.behaviorTick++ % this.activatedPriority == 0) // Leaf - Plazma - Add missing Pufferfish configurations
             this.getBrain().tick(level, this);
             profiler.pop();
             profiler.push("happyGhastActivityUpdate");
diff --git a/net/minecraft/world/entity/animal/nautilus/Nautilus.java b/net/minecraft/world/entity/animal/nautilus/Nautilus.java
--- a/net/minecraft/world/entity/animal/nautilus/Nautilus.java
+++ b/net/minecraft/world/entity/animal/nautilus/Nautilus.java
@@ -100,6 +100,7 @@ public class Nautilus extends AbstractNautilus {
         return baby;
     }
 
+    private int behaviorTick; // Leaf - Plazma - Add missing Pufferfish configurations
     @Override
     public EntityDimensions getDefaultDimensions(final Pose pose) {
         return this.isBaby() ? BABY_DIMENSIONS : super.getDefaultDimensions(pose);
@@ -109,6 +110,7 @@ public class Nautilus extends AbstractNautilus {
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("nautilusBrain");
+        if (this.behaviorTick++ % this.activatedPriority == 0) // Leaf - Plazma - Add missing Pufferfish configurations
         this.getBrain().tick(level, this);
         profiler.pop();
         profiler.push("nautilusActivityUpdate");
diff --git a/net/minecraft/world/entity/animal/nautilus/ZombieNautilus.java b/net/minecraft/world/entity/animal/nautilus/ZombieNautilus.java
--- a/net/minecraft/world/entity/animal/nautilus/ZombieNautilus.java
+++ b/net/minecraft/world/entity/animal/nautilus/ZombieNautilus.java
@@ -98,10 +98,12 @@ public class ZombieNautilus extends AbstractNautilus {
         return (Brain<ZombieNautilus>)super.getBrain();
     }
 
+    private int behaviorTick; // Leaf - Plazma - Add missing Pufferfish configurations
     @Override
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("zombieNautilusBrain");
+        if (this.behaviorTick++ % this.activatedPriority == 0) // Leaf - Plazma - Add missing Pufferfish configurations
         this.getBrain().tick(level, this);
         profiler.pop();
         profiler.push("zombieNautilusActivityUpdate");
diff --git a/net/minecraft/world/entity/animal/sniffer/Sniffer.java b/net/minecraft/world/entity/animal/sniffer/Sniffer.java
--- a/net/minecraft/world/entity/animal/sniffer/Sniffer.java
+++ b/net/minecraft/world/entity/animal/sniffer/Sniffer.java
@@ -493,11 +493,12 @@ public class Sniffer extends Animal {
         return (Brain<Sniffer>)super.getBrain();
     }
 
+    private int behaviorTick; // Leaf - Plazma - Add missing Pufferfish configurations
     @Override
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("snifferBrain");
-        if (getRider() == null || !this.isControllable()) // Purpur - only use brain if no rider
+        if ((getRider() == null || !this.isControllable()) && this.behaviorTick++ % this.activatedPriority == 0) // Purpur - only use brain if no rider // Leaf - Plazma - Add missing Pufferfish configurations
         this.getBrain().tick(level, this);
         profiler.popPush("snifferActivityUpdate");
         SnifferAi.updateActivity(this);
diff --git a/net/minecraft/world/entity/monster/Zoglin.java b/net/minecraft/world/entity/monster/Zoglin.java
--- a/net/minecraft/world/entity/monster/Zoglin.java
+++ b/net/minecraft/world/entity/monster/Zoglin.java
@@ -275,11 +275,12 @@ public class Zoglin extends Monster implements HoglinBase {
         this.setAggressive(this.brain.hasMemoryValue(MemoryModuleType.ATTACK_TARGET));
     }
 
+    private int behaviorTick; // Leaf - Plazma - Add missing Pufferfish configurations
     @Override
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("zoglinBrain");
-        if (getRider() == null || !this.isControllable()) // Purpur - only use brain if no rider
+        if ((getRider() == null || !this.isControllable()) && this.behaviorTick++ % this.activatedPriority == 0) // Purpur - only use brain if no rider // Leaf - Plazma - Add missing Pufferfish configurations
         this.getBrain().tick(level, this);
         profiler.pop();
         this.updateActivity();
diff --git a/net/minecraft/world/entity/monster/breeze/Breeze.java b/net/minecraft/world/entity/monster/breeze/Breeze.java
--- a/net/minecraft/world/entity/monster/breeze/Breeze.java
+++ b/net/minecraft/world/entity/monster/breeze/Breeze.java
@@ -296,11 +296,12 @@ public class Breeze extends Monster {
         return target.closerThan(ourPosition, 4.0, 10.0);
     }
 
+    private int behaviorTick; // Leaf - Plazma - Add missing Pufferfish configurations
     @Override
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("breezeBrain");
-        if (getRider() == null || !this.isControllable()) // Purpur - only use brain if no rider
+        if ((getRider() == null || !this.isControllable()) && this.behaviorTick++ % this.activatedPriority == 0) // Purpur - only use brain if no rider // Leaf - Plazma - Add missing Pufferfish configurations
         this.getBrain().tick(level, this);
         profiler.popPush("breezeActivityUpdate");
         BreezeAi.updateActivity(this);
diff --git a/net/minecraft/world/entity/monster/creaking/Creaking.java b/net/minecraft/world/entity/monster/creaking/Creaking.java
--- a/net/minecraft/world/entity/monster/creaking/Creaking.java
+++ b/net/minecraft/world/entity/monster/creaking/Creaking.java
@@ -235,10 +235,12 @@ public class Creaking extends Monster {
         return (Brain<Creaking>)super.getBrain();
     }
 
+    private int behaviorTick; // Leaf - Plazma - Add missing Pufferfish configurations
     @Override
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("creakingBrain");
+        if (this.behaviorTick++ % this.activatedPriority == 0) // Leaf - Plazma - Add missing Pufferfish configurations
         this.getBrain().tick((ServerLevel)this.level(), this);
         profiler.pop();
         CreakingAi.updateActivity(this);
diff --git a/net/minecraft/world/entity/monster/piglin/PiglinBrute.java b/net/minecraft/world/entity/monster/piglin/PiglinBrute.java
--- a/net/minecraft/world/entity/monster/piglin/PiglinBrute.java
+++ b/net/minecraft/world/entity/monster/piglin/PiglinBrute.java
@@ -138,11 +138,12 @@ public class PiglinBrute extends AbstractPiglin {
         return itemStack.is(Items.GOLDEN_AXE) && super.wantsToPickUp(level, itemStack);
     }
 
+    private int behaviorTick; // Leaf - Plazma - Add missing Pufferfish configurations
     @Override
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("piglinBruteBrain");
-        if (getRider() == null || this.isControllable()) // Purpur - only use brain if no rider
+        if ((getRider() == null || this.isControllable()) && this.behaviorTick++ % this.activatedPriority == 0) // Purpur - only use brain if no rider // Leaf - Plazma - Add missing Pufferfish configurations
         this.getBrain().tick(level, this);
         profiler.pop();
         PiglinBruteAi.updateActivity(this);
